chore(agw): Python dependency upgrades

MoritzThomasHuebner opened the pull request:
Fixes #13426

In pairing with @wolfseb 

## Summary

Done:
- Upgraded Python versions where possible. We used `pipdeptree` to analyze packages and dependencies.
- Temporary change to upgrade `fpm` during provisioning. This is necessary to build `cryptography`.
- Some packages (such as docker) are not built by `pydep`, since a `python3-docker` exists in `http://archive.ubuntu.com/ubuntu/`.
- Updated lockfile


## Test Plan

-  Run `fab release package:destroy_vm=True` on the host VM to trigger new magma and dependency `*.deb` files to be built.
- Install local `*.deb` files, and any other dependencies that have not been newly packaged from the artifactory.
- Remove `/etc/magma`.
- Install `magma*.deb` file.
- Spin up `magma_test` and `magma_trfserver` 
- Run S1AP precommit and integration tests.
- Run `build-all` script on a fork: https://github.com/MoritzThomasHuebner/magma/runs/7759753398?check_suite_focus=true
  - For this run we needed to make a temporary change that disables the check that we are on the origin repo.
  - The check shows that all packages are actually being built.
- Run Bazel build and test workflow on fork: https://github.com/MoritzThomasHuebner/magma/actions/runs/2830383274
  - Ensures versions in `bazel/external/requirements.in` do not cause any version conflicts

## Additional Information

There are a number of packages we were not able to upgrade:
- `fpm` requires that there is a `setup.py`, thus some of the newest versions of `pyparsing`, `jsonschema`, `pyroute2` can no longer be built since these packages now utilize other build methods.
- `setuptools`: The version `49.6.0` is hardcoded in some places.
- `docker`: Ubuntu focal is not compatible with `docker>=5`.
  - `requests`: The latest versions require `docker>=5`.
- `flask` is restricted to `1.1.4`.
  - This restricts `click`, `itsdangerous`, `Jinja2`, and `markupsafe`.
- `redis` is restricted to `3.5.3`.
- `cryptography` is restricted to `3.2.1` because newer versions cannot be built with `fpm`.
- `PyYAML` is restricted to `<6.0`


- [ ] This change is backwards-breaking

MoritzThomasHuebner committed:
chore(agw): Updating python dependencies

- Upgraded versions in `lte/gateway/setup.py` and `orc8r/gateway/setup.py` where possible.
- Added `--python-pip /home/vagrant/build/python/bin/pip` to the `fpm` command in `pydep`. This option ensures that we are not using `easy_install` inadvertently.
- Added a step to upgrade `fpm` upon VM provisioning. This step should be removed after the next preburn is created. The `fpm` upgrade is necessary to build a `.deb` file for `cryptography`.
- Updated the `magma.lockfile.ubuntu` by running `fab release package`.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Updating python dependencies

- Added pylint exception. The linter erroneously flags this line otherwise.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Updating python dependencies

- Updated bazel files.
- Also replaced `systemd` with `systemd-python` in bazel files since these are different packages

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Updating python dependencies

- Updated bazel files.
- Also replaced `systemd` with `systemd-python` in bazel files since these are different packages

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

alexzurbonsen committed:
downgrade pyyaml

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen committed:
bazel: downgrade pyyaml and generate requirements.txt changes

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

sebathomas commented on .github/workflows/bazel.yml:
```diff
@@ -142,9 +142,9 @@ jobs:
-      needs.path_filter.outputs.files_changed == 'true'
+#    if: |
+#      (github.event_name == 'schedule' && github.ref == 'refs/heads/master') ||
+#      needs.path_filter.outputs.files_changed == 'true'
```
I think these commented-out filters need to be reverted?

MoritzThomasHuebner replied:
Done, we commented these out temporarily to run the workflow on my fork.

sebathomas commented on .github/workflows/build_all.yml:
```diff
@@ -101,7 +101,7 @@ jobs:
           python ci-scripts/helm_repo_rotation.py
   agw-build:
-    if: github.event_name == 'push' && github.repository_owner == 'magma'
+#    if: github.event_name == 'push' && github.repository_owner == 'magma'
```
Same here.

MoritzThomasHuebner replied:
Done

sebathomas commented on lte/gateway/deploy/roles/dev_common/tasks/main.yml:
```diff
@@ -158,6 +158,12 @@
   shell: gem install fpm
   when: preburn
```
Nit: Usually, we just set the step to "full_provision" with a TODO rather than duplicating it.

MoritzThomasHuebner replied:
Done

sebathomas reviewed (changes requested):
This is looking good, except for the commented-out code. Thanks for finally getting these updates working!

By the way, I think "chore(agw)" would be a more fitting tag than "ci".

MoritzThomasHuebner committed:
chore(agw): Requested changes from github

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng reviewed (approved):
bazel + pipelined lgtm

nstng reviewed (changes requested):
Will block this intentionally because we have a version issue with `sentry-sdk` in the 1.8 RC that we might want to fix less invasive and merge first for backporting.
See https://github.com/magma/magma/issues/13645

jheidbrink commented on bazel/external/requirements.in:
```diff
@@ -1,43 +1,44 @@
+cryptography==3.2.1
 #  h2>=3,<4 is requirement of aioh2 (loaded via bazel)
-h2>=3,<4
+h2==3.2.0
```
What is the reason for pinning most of the versions in `requirements.in`? As far as I understand the requirements.in, it is supposed to contain version constraints, and then `pip-compile` generates the `requirements.txt` which has exact pinnings. The `requirements.txt` will then be used by `pip` to install the exactly pinned versions.

This line for example would previously tell `pip-compile` to not upgrade to a version `>= 4.0`. With the line changed to `h2==3.2.0`, `pip-compile` no longer knows about this constraint.

Without the exact pinning, I can use other package repositories like nixpkgs to install Python dependencies. This stops working when we exactly pin the dependencies, because that very specific combination of packages usually isn't available there.

jheidbrink replied:
Related question: Why do we convert some of the `>=` lines to `==`, while others stay `>=` constraints?

sebathomas replied:
I some cases (e.g. jinja2), things break if you allow `>=` and pip-compile won't be able to resolve it. Not sure about h2 in particular.

jheidbrink replied:
Ah, i see, thanks!

MoritzThomasHuebner replied:
Thanks, I am just copying below what I sent privately on slack:

"I had another look into the [requirements.in](http://requirements.in/) file and how we compile this to a requirements.txt. One thing I have noticed today is that the output of pip-compile depends on what is already in the requirements.txt. In fact, if you delete the requirements.txt on the current state of master and try to build it again, it will fail (see screenshot). To resolve the version conflict in the screenshot, I manually set flask<2.1.0.  pip-compile apparently only changes versions in the requirements.txt if it is forced to. If the existing versions fulfil the [requirements.in](http://requirements.in/), pip-compile will not upgrade versions.
With all of that in mind, I changed the PR so that we only use fixed versions where they were already in place. I incremented all of the fixed versions, added the constraint on flask, and copied the sentry-sdk constraint from the setup.py file. Additionally, I changed systemd to systemd-python. These are actually different packages with similar interfaces. systemd-python is what is actually required in magma.
Finally, I removed the requirements.txt , built it again from scratch, and added the changes to the PR."

jheidbrink replied:
Yes, that seems to be in accordance with https://github.com/jazzband/pip-tools#example-usage-for-pip-compile:
> Note: If pip-compile finds an existing requirements.txt file that fulfils the dependencies then no changes will be made, even if updates are available. To compile from scratch, first delete the existing requirements.txt file, or see [Updating requirements](https://github.com/jazzband/pip-tools#updating-requirements) for alternative approaches.

alexzurbonsen replied:
> What is the reason for pinning most of the versions in `requirements.in`? 

That originated from a conversation between @LKreutzer and me, who brought https://github.com/magma/magma/issues/11743 to my attention. The second bullet point, according to our interpretation, implied that one wants to make sure, that same versions are used independently of whether you rely on setup.py or requirements.in.

But just as an explanation, I guess its fine without that as well.

jheidbrink replied:
Good point, ensuring that installed versions are the same when using `pip install` vs building & installing the `.deb` package is a good argument for pinning in `requirements.in`.

jheidbrink replied:
But I think the current change where we keep the pinning for packages that were already pinned is fine for now.

nstng commented on lte/gateway/python/setup.py:
```diff
@@ -137,9 +137,9 @@
             # Keep grpcio and grpcio-tools on same version for now
```
Is this comment outdated? To me it looks like grpcio-tools>=1.46.3 is used here and grpcio>=1.47.0 in orc8r/setup.py

sebathomas replied:
Was there a reason to use different versions? I assume we know that it works due to the tests but I would still prefer to keep them the same.

MoritzThomasHuebner replied:
I made the versions consistent

nstng commented on lte/gateway/python/setup.py:
```diff
@@ -137,9 +137,9 @@
             # Keep grpcio and grpcio-tools on same version for now
             # If you update this version here, you probably also want to
```
This seems outdated - I do not see grpc setup in lte/gateway/python/Makefile

MoritzThomasHuebner replied:
I removed the comment

jheidbrink commented on bazel/external/requirements.in:
```diff
@@ -1,43 +1,44 @@
+cryptography==3.2.1
 #  h2>=3,<4 is requirement of aioh2 (loaded via bazel)
-h2>=3,<4
+h2==3.2.0
```
What is the reason for exactly pinning most packages in `requirements.in`? As far as I understand the mechanism, `pip-compile` will already create a `requirements.txt` with exact pinnings, and pip will use `requirements.txt` to install exactly pinned packages. The `requirements.in` can then be used to define versions constraints for `pip-compile`.

In this example, the previous version of the line would tell `pip-compile` not to upgrade to `h2 >= 4.0`. With the new line `h2==3.2.0`, `pip-compile` no longer has that information.

That scheme also allows different package managers that don't necessarily have all the packages from pypi. For example a developer could choose to install Python dependencies from nixpkgs on their machine. With the exact pinnings in `requirements.in` this doesn't work anymore as nixpkgs won't have that exact combination of packages available.

jheidbrink replied:
Sorry, duplicate comment - for some reason Github didn't show my previous comment and I thought it was gone...

MoritzThomasHuebner committed:
chore(agw): More Requested changes from github

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): More Requested changes from github

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Fixing version issues with click

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Replace systemd with systemd_python

This change makes bazel consistent with the setup.py files. The two packages have similar interfaces but are different and cannot just be interchanged. The use of `systemd.daemon.notify` in `sdwatchdog.py` also indicates that `systemd_python` is the intended package one should use.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Revert to using easy_install

Using pip to download some packages for fpm fails currently since fpm uses adds the `--no-binary :all:` flag to pip download.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Remove fpm upgrade during provision

fpm is now already on the latest version when you spin up the VM.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng reviewed (approved):
bazel + pipelined lgtm
